snuba/optimize.py

```diff
@@ -36,20 +37,43 @@ def run_optimize(
     storage: ReadableTableStorage,
     database: str,
     before: Optional[datetime] = None,
-    ignore_cutoff: bool = False,
     parallel: int = 1,
+    cutoff_time: Optional[datetime] = None,
     clickhouse_host: Optional[str] = None,
+    optimize_partition_tracker: Optional[OptimizedPartitionTracker] = None,
 ) -> int:
     start = time.time()
     schema = storage.get_schema()
     assert isinstance(schema, TableSchema)
     table = schema.get_local_table_name()
     database = storage.get_cluster().get_database()
-
     parts = get_partitions_to_optimize(clickhouse, storage, database, table, before)
+
+    if optimize_partition_tracker:
+        completed_partitions = optimize_partition_tracker.get_completed_partitions()
+        parts = (
+            [part for part in parts if part.name not in completed_partitions]
+            if completed_partitions
+            else parts
+        )
+
+    if not len(parts):
+        logger.info("No partitions to optimize")
+        return 0
+
     optimize_partition_runner(
-        clickhouse, database, table, parts, ignore_cutoff, parallel, clickhouse_host
+        clickhouse=clickhouse,
+        database=database,
+        table=table,
+        parts=parts,
+        parallel=parallel,
+        cutoff_time=cutoff_time,
+        optimize_partition_tracker=optimize_partition_tracker,
+        clickhouse_host=clickhouse_host,
     )
+
+    if optimize_partition_tracker:
+        optimize_partition_tracker.remove_all_partitions()
```

Comment:
How precise is the guarantee that Redis will get rid of a key at the TTL? If that is guaranteed all good. But if that means that Redis will keep the key "at least" up to the TTL, we may end up restarting the job the following day in case of exception or crash with the state still present.
A way to be 100% certain is to partition the state by day. See the comment below. if your state contains the date, you are sure never to read the state of the day before.

Reply:
Date is part of the key now